[Index Management] Fix @custom component template creation

kapral18 opened the pull request:
Addresses #237916.

- Fix `/component_templates/{name}/datastreams` route to return empty when the component template is not referenced by any index template, preventing unintended apply-mappings and the rollover modal for brand-new `@custom` component templates
- add test coverage for `component_templates` routes folder
- refactor testing setup

## Manual test setup

1. Run latest es/kibana locally from kibana repo source
2. In DevTools:
```graphql
PUT _index_template/foo-template
{
  "index_patterns": ["foo-ds-*"],
  "data_stream": {},
  "template": {
    "mappings": {
      "properties": {
        "message": { "type": "text" },
        "timestamp": { "type": "date" }
      }
    }
  }
}

PUT _data_stream/foo-ds-1

# Introduce a conflict mapping in index-template
PUT _index_template/foo-template
{
  "index_patterns": ["foo-ds-*"],
  "data_stream": {},
  "template": {
    "mappings": {
      "properties": {
        "message": { "type": "boolean" },
        "timestamp": { "type": "date" }
      }
    }
  }
}
```
 4. Go to `/app/management/data/index_management/component_templates`
 5. Create a new template that is not referenced by any `index_template` and has a `@custom` suffix, ex. `random-1234@custom`
 6. Verify the behavior: Before this PR change a modal asking to rollover conflicting data-stream foo-ds-1 will appear, and after this PR no modal will appear.
 
!Important Do not click `apply now and rollover` in the modal as it will mess up setup to verify before/after fix.

## Release note

- Fixed a bug in component template creation flow where a new component template with `@custom` suffix in name would lead to updating mappings of all unrelated data-streams and a popup would appear asking to rollover conflicting ones.


<!--ONMERGE {"backportTargets":["8.18","8.19","9.1","9.2"]} ONMERGE-->

kapral18 committed:
[Index Management] Fix @custom component template creation

Addresses #237916.

- Fix `/component_templates/{name}/datastreams` route to return empty
when the component template is not referenced by any index template,
preventing unintended apply-mappings and the rollover modal for
brand-new `@custom` component templates
- add test coverage for `component_templates` routes folder
- refactor testing setup

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/348521)
* Commit: c869ae45c22ce85f831880edab76d22ddae5bc62

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/348521#0199c165-d772-4e7a-ae09-091597d526c4)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/348521","number":348521,"commit":"c869ae45c22ce85f831880edab76d22ddae5bc62"}],"number":348521}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

ElenaStoeva commented:
Hi @kapral18, I tested locally and I don't see the modal. However, I also don't see it without the changes from this PR so I might be doing something wrong - let's meet tomorrow and discuss this.

ElenaStoeva reviewed (approved):
Thanks for updating the instructions! Tested locally and verified that:
- The modal is not displayed when the component template is not referenced by a conflicting index template
- The modal is not displayed when the component is referenced by a non-conflicting index template
- The modal is correctly displayed when the component template is referenced by a conflicting index template

Code changes also lgtm. Thanks for adding tests!

I'm approving to unblock the PR, but I wonder if we should test this behavior somehow in a follow-up, maybe with functional tests. Wdyt?

kapral18 committed:
refactor: Rename datastream rollover to mappings update
- Renamed the `useDatastreamsRollover` hook to `useUpdateAssociatedDatastreamsMappings`.
- The corresponding test file and modal component have also been renamed to reflect this change.
- The functionality previously described as "rolling over linked datastreams" is now referred to as "updating mappings from template for associated data streams."
- The `showDatastreamRolloverModal` function is now `updateAssociatedDatastreamsMappings`, and its description clarifies that a modal is prompted for rollover only if mapping updates fail due to incompatible changes.

kapral18 committed:
ci: fix windows path limitation issue
